pylint/lint.py

```diff
@@ -1054,31 +1082,57 @@ def _do_check(self, files_or_modules):
             if interfaces.implements(checker, interfaces.IAstroidChecker):
                 walker.add_checker(checker)
         # build ast and check modules or packages
-        for descr in self.expand_files(files_or_modules):
-            modname, filepath, is_arg = descr["name"], descr["path"], descr["isarg"]
-            if not self.should_analyze_file(modname, filepath, is_argument=is_arg):
-                continue
+        if self.config.from_stdin:
+            assert len(files_or_modules) == 1
+            filepath = files_or_modules[0]
+            try:
+                # Note that modpath_from_file does not really perform an
+                # __import__ as one may expect, because ImportError
+                # exceptions are explictly catched here.
+                modname = ".".join(modutils.modpath_from_file(filepath))
+            except ImportError:
+                modname = os.path.splitext(os.path.basename(filepath))[0]
 
             self.set_current_module(modname, filepath)
+
             # get the module representation
-            ast_node = self.get_ast(filepath, modname)
-            if ast_node is None:
-                continue
-            # XXX to be correct we need to keep module_msgs_state for every
-            # analyzed module (the problem stands with localized messages which
-            # are only detected in the .close step)
-            self.file_state = utils.FileState(descr["basename"])
-            self._ignore_file = False
-            # fix the current file (if the source file was not available or
-            # if it's actually a c extension)
-            self.current_file = ast_node.file  # pylint: disable=maybe-no-member
-            self.check_astroid_module(ast_node, walker, rawcheckers, tokencheckers)
-            # warn about spurious inline messages handling
-            spurious_messages = self.file_state.iter_spurious_suppression_messages(
-                self.msgs_store
-            )
-            for msgid, line, args in spurious_messages:
-                self.add_message(msgid, line, None, args)
+            ast_node = _ast_from_string(_read_stdin(), filepath, modname)
+
+            if ast_node is not None:
+                self.file_state = utils.FileState(filepath)
+                self.check_astroid_module(ast_node, walker, rawcheckers, tokencheckers)
+                # warn about spurious inline messages handling
```

Comment:
This block seems duplicated with the normal flow, maybe move it after the whole `if` branch?

Reply:
Moving it after whole `if self.config.from_stdin` branch does not work IMO. I think it is not worth the effort to avoid this code duplication, especially since only a couple of lines of code are duplicated.